{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cada1e0e",
   "metadata": {},
   "source": [
    "## Read TAIR GFF file (into a pandas dataframe)\n",
    "\n",
    "Source URL: https://www.arabidopsis.org/download/index-auto.jsp?dir=%2Fdownload_files%2FGenes%2FAraport11_genome_release"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37f45946",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "tair_df = pd.read_csv('Araport11_GFF3_genes_transposons.current.gff.gz', \n",
    "                      sep='\\t', \n",
    "                      comment='#', \n",
    "                      encoding='cp1252', \n",
    "                      header=None, \n",
    "                     compression='gzip')\n",
    "\n",
    "tair_df.columns = ['chromosome', 'data_source', 'type', 'start', 'end', 'score', 'strand', 'frame', 'features']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19ca4e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def string_to_dict(input_string):\n",
    "    '''Function to convert feature string into a dictionary.'''\n",
    "    try:\n",
    "        # Split the input string into a list of key-value pairs using semicolon as the delimiter\n",
    "        feature_list = input_string.split(';')\n",
    "\n",
    "        # Filter the list to only include items that contain an equal sign ('=')\n",
    "        feature_list = [x for x in feature_list if '=' in x]\n",
    "\n",
    "        # Create a dictionary by splitting each key-value pair at the first equal sign encountered\n",
    "        feature_dict = dict(x.split('=', 1) for x in feature_list)\n",
    "    except:\n",
    "        # If there's an exception (e.g., invalid input), print the original input string and return an empty dictionary\n",
    "        print(input_string)\n",
    "        feature_dict = dict()\n",
    "    \n",
    "    # Return the resulting dictionary\n",
    "    return feature_dict\n",
    "\n",
    "# The code provided assumes there's a variable 'inp_str' containing the input string\n",
    "# You should replace 'inp_str' with an actual input string before calling the function.\n",
    "# Example usage:\n",
    "# input_string = \"key1=value1;key2=value2;key3=value3\"\n",
    "# result = string_to_dict(input_string)\n",
    "# This will convert the input string into a dictionary and store it in the 'result' variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe24b5a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nan\n",
      "nan\n"
     ]
    }
   ],
   "source": [
    "tair_df['features_dict'] = tair_df.features.apply(string_to_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec645a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# string_to_dict('ID=AT1G01010.1;Parent=AT1G01010;Name=AT1G01010.1;Note=NAC domain containing protein 1;conf_class=2;symbol=NAC001;full_name=NAC domain containing protein 1;computational_description=NAC domain containing protein 1;conf_rating=****;gene=2200934,UniProt=Q0WV96;curator_summary=Member of the NAC domain containing family of plant specific transcriptional regulators.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dcd063f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         {'ID': 'AT1G01010', 'Name': 'AT1G01010', 'Note...\n",
       "1         {'ID': 'AT1G01010.1', 'Parent': 'AT1G01010', '...\n",
       "2         {'ID': 'AT1G01010:CDS:1', 'Parent': 'AT1G01010...\n",
       "3         {'ID': 'AT1G01010:CDS:2', 'Parent': 'AT1G01010...\n",
       "4         {'ID': 'AT1G01010:CDS:3', 'Parent': 'AT1G01010...\n",
       "                                ...                        \n",
       "927370    {'ID': 'ATMG01380:exon:1', 'Parent': 'ATMG0138...\n",
       "927371    {'ID': 'ATMG01380.1', 'Parent': 'ATMG01380', '...\n",
       "927372    {'ID': 'ATMG01390', 'id2': 'gene-rrn18', 'Name...\n",
       "927373    {'ID': 'ATMG01390:exon:1', 'Parent': 'ATMG0139...\n",
       "927374    {'ID': 'ATMG01390.1', 'Parent': 'ATMG01390', '...\n",
       "Name: features_dict, Length: 927375, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair_df.features_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df881abd",
   "metadata": {},
   "source": [
    "# Filter TAIR entries\n",
    "\n",
    "#### Filtering entries based on column values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a5fcc0f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>features</th>\n",
       "      <th>features_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>3760.0</td>\n",
       "      <td>5630.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01010.1-Protein;Name=AT1G01010.1;Derive...</td>\n",
       "      <td>{'ID': 'AT1G01010.1-Protein', 'Name': 'AT1G010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>23519.0</td>\n",
       "      <td>31079.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01040.1-Protein;Name=AT1G01040.1;Derive...</td>\n",
       "      <td>{'ID': 'AT1G01040.1-Protein', 'Name': 'AT1G010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>23519.0</td>\n",
       "      <td>31079.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01040.2-Protein;Name=AT1G01040.2;Derive...</td>\n",
       "      <td>{'ID': 'AT1G01040.2-Protein', 'Name': 'AT1G010...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>53022.0</td>\n",
       "      <td>54494.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01110.1-Protein;Name=AT1G01110.1;Derive...</td>\n",
       "      <td>{'ID': 'AT1G01110.1-Protein', 'Name': 'AT1G011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>52239.0</td>\n",
       "      <td>54494.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01110.2-Protein;Name=AT1G01110.2;Derive...</td>\n",
       "      <td>{'ID': 'AT1G01110.2-Protein', 'Name': 'AT1G011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927135</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>139856.0</td>\n",
       "      <td>140650.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01230.1-Protein;Name=ATCG01230.1;Derive...</td>\n",
       "      <td>{'ID': 'ATCG01230.1-Protein', 'Name': 'ATCG012...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927140</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>140704.0</td>\n",
       "      <td>141171.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01240.1-Protein;Name=ATCG01240.1;Derive...</td>\n",
       "      <td>{'ID': 'ATCG01240.1-Protein', 'Name': 'ATCG012...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927147</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>141854.0</td>\n",
       "      <td>143708.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01250.1-Protein;Name=ATCG01250.1;Derive...</td>\n",
       "      <td>{'ID': 'ATCG01250.1-Protein', 'Name': 'ATCG012...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927168</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>152506.0</td>\n",
       "      <td>152787.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01300.1-Protein;Name=ATCG01300.1;Derive...</td>\n",
       "      <td>{'ID': 'ATCG01300.1-Protein', 'Name': 'ATCG013...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927175</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>protein</td>\n",
       "      <td>152806.0</td>\n",
       "      <td>154312.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01310.1-Protein;Name=ATCG01310.1;Derive...</td>\n",
       "      <td>{'ID': 'ATCG01310.1-Protein', 'Name': 'ATCG013...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24256 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome data_source     type     start       end score strand frame  \\\n",
       "15           Chr1   Araport11  protein    3760.0    5630.0     .      +     .   \n",
       "200          Chr1   Araport11  protein   23519.0   31079.0     .      +     .   \n",
       "244          Chr1   Araport11  protein   23519.0   31079.0     .      +     .   \n",
       "603          Chr1   Araport11  protein   53022.0   54494.0     .      +     .   \n",
       "617          Chr1   Araport11  protein   52239.0   54494.0     .      +     .   \n",
       "...           ...         ...      ...       ...       ...   ...    ...   ...   \n",
       "927135       ChrC   Araport11  protein  139856.0  140650.0     .      +     .   \n",
       "927140       ChrC   Araport11  protein  140704.0  141171.0     .      +     .   \n",
       "927147       ChrC   Araport11  protein  141854.0  143708.0     .      +     .   \n",
       "927168       ChrC   Araport11  protein  152506.0  152787.0     .      +     .   \n",
       "927175       ChrC   Araport11  protein  152806.0  154312.0     .      +     .   \n",
       "\n",
       "                                                 features  \\\n",
       "15      ID=AT1G01010.1-Protein;Name=AT1G01010.1;Derive...   \n",
       "200     ID=AT1G01040.1-Protein;Name=AT1G01040.1;Derive...   \n",
       "244     ID=AT1G01040.2-Protein;Name=AT1G01040.2;Derive...   \n",
       "603     ID=AT1G01110.1-Protein;Name=AT1G01110.1;Derive...   \n",
       "617     ID=AT1G01110.2-Protein;Name=AT1G01110.2;Derive...   \n",
       "...                                                   ...   \n",
       "927135  ID=ATCG01230.1-Protein;Name=ATCG01230.1;Derive...   \n",
       "927140  ID=ATCG01240.1-Protein;Name=ATCG01240.1;Derive...   \n",
       "927147  ID=ATCG01250.1-Protein;Name=ATCG01250.1;Derive...   \n",
       "927168  ID=ATCG01300.1-Protein;Name=ATCG01300.1;Derive...   \n",
       "927175  ID=ATCG01310.1-Protein;Name=ATCG01310.1;Derive...   \n",
       "\n",
       "                                            features_dict  \n",
       "15      {'ID': 'AT1G01010.1-Protein', 'Name': 'AT1G010...  \n",
       "200     {'ID': 'AT1G01040.1-Protein', 'Name': 'AT1G010...  \n",
       "244     {'ID': 'AT1G01040.2-Protein', 'Name': 'AT1G010...  \n",
       "603     {'ID': 'AT1G01110.1-Protein', 'Name': 'AT1G011...  \n",
       "617     {'ID': 'AT1G01110.2-Protein', 'Name': 'AT1G011...  \n",
       "...                                                   ...  \n",
       "927135  {'ID': 'ATCG01230.1-Protein', 'Name': 'ATCG012...  \n",
       "927140  {'ID': 'ATCG01240.1-Protein', 'Name': 'ATCG012...  \n",
       "927147  {'ID': 'ATCG01250.1-Protein', 'Name': 'ATCG012...  \n",
       "927168  {'ID': 'ATCG01300.1-Protein', 'Name': 'ATCG013...  \n",
       "927175  {'ID': 'ATCG01310.1-Protein', 'Name': 'ATCG013...  \n",
       "\n",
       "[24256 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair_df[(tair_df.type == 'protein') & (tair_df.strand == '+')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f35ccf77",
   "metadata": {},
   "source": [
    "# Filter TAIR entries based on feature(s)\n",
    "\n",
    "#### Filtering entries that have five star (\\*\\*\\*\\*\\*) confidence rating."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6670bada",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>features</th>\n",
       "      <th>features_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>38752.0</td>\n",
       "      <td>40945.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01070.2;Parent=AT1G01070;Name=AT1G01070...</td>\n",
       "      <td>{'ID': 'AT1G01070.2', 'Parent': 'AT1G01070', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>52061.0</td>\n",
       "      <td>54689.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01110.1;Parent=AT1G01110;Name=AT1G01110...</td>\n",
       "      <td>{'ID': 'AT1G01110.1', 'Parent': 'AT1G01110', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>57164.0</td>\n",
       "      <td>59215.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01120.1;Parent=AT1G01120;Name=AT1G01120...</td>\n",
       "      <td>{'ID': 'AT1G01120.1', 'Parent': 'AT1G01120', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>820</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>75390.0</td>\n",
       "      <td>76845.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01180.1;Parent=AT1G01180;Name=AT1G01180...</td>\n",
       "      <td>{'ID': 'AT1G01180.1', 'Parent': 'AT1G01180', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>86486.0</td>\n",
       "      <td>88409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01200.1;Parent=AT1G01200;Name=AT1G01200...</td>\n",
       "      <td>{'ID': 'AT1G01200.1', 'Parent': 'AT1G01200', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927130</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>139856.0</td>\n",
       "      <td>140650.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01230.1;Parent=ATCG01230;Name=ATCG01230...</td>\n",
       "      <td>{'ID': 'ATCG01230.1', 'Parent': 'ATCG01230', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927137</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>140704.0</td>\n",
       "      <td>141171.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01240.1;Parent=ATCG01240;Name=ATCG01240...</td>\n",
       "      <td>{'ID': 'ATCG01240.1', 'Parent': 'ATCG01240', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927152</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>144921.0</td>\n",
       "      <td>145154.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01270.1;Parent=ATCG01270;Name=ATCG01270...</td>\n",
       "      <td>{'ID': 'ATCG01270.1', 'Parent': 'ATCG01270', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927157</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>145291.0</td>\n",
       "      <td>152175.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01280.1;Parent=ATCG01280;Name=ATCG01280...</td>\n",
       "      <td>{'ID': 'ATCG01280.1', 'Parent': 'ATCG01280', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927165</th>\n",
       "      <td>ChrC</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>152506.0</td>\n",
       "      <td>152787.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ATCG01300.1;Parent=ATCG01300;Name=ATCG01300...</td>\n",
       "      <td>{'ID': 'ATCG01300.1', 'Parent': 'ATCG01300', '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8894 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome data_source  type     start       end score strand frame  \\\n",
       "481          Chr1   Araport11  mRNA   38752.0   40945.0     .      -     .   \n",
       "593          Chr1   Araport11  mRNA   52061.0   54689.0     .      +     .   \n",
       "620          Chr1   Araport11  mRNA   57164.0   59215.0     .      -     .   \n",
       "820          Chr1   Araport11  mRNA   75390.0   76845.0     .      +     .   \n",
       "846          Chr1   Araport11  mRNA   86486.0   88409.0     .      -     .   \n",
       "...           ...         ...   ...       ...       ...   ...    ...   ...   \n",
       "927130       ChrC   Araport11  mRNA  139856.0  140650.0     .      +     .   \n",
       "927137       ChrC   Araport11  mRNA  140704.0  141171.0     .      +     .   \n",
       "927152       ChrC   Araport11  mRNA  144921.0  145154.0     .      -     .   \n",
       "927157       ChrC   Araport11  mRNA  145291.0  152175.0     .      -     .   \n",
       "927165       ChrC   Araport11  mRNA  152506.0  152787.0     .      +     .   \n",
       "\n",
       "                                                 features  \\\n",
       "481     ID=AT1G01070.2;Parent=AT1G01070;Name=AT1G01070...   \n",
       "593     ID=AT1G01110.1;Parent=AT1G01110;Name=AT1G01110...   \n",
       "620     ID=AT1G01120.1;Parent=AT1G01120;Name=AT1G01120...   \n",
       "820     ID=AT1G01180.1;Parent=AT1G01180;Name=AT1G01180...   \n",
       "846     ID=AT1G01200.1;Parent=AT1G01200;Name=AT1G01200...   \n",
       "...                                                   ...   \n",
       "927130  ID=ATCG01230.1;Parent=ATCG01230;Name=ATCG01230...   \n",
       "927137  ID=ATCG01240.1;Parent=ATCG01240;Name=ATCG01240...   \n",
       "927152  ID=ATCG01270.1;Parent=ATCG01270;Name=ATCG01270...   \n",
       "927157  ID=ATCG01280.1;Parent=ATCG01280;Name=ATCG01280...   \n",
       "927165  ID=ATCG01300.1;Parent=ATCG01300;Name=ATCG01300...   \n",
       "\n",
       "                                            features_dict  \n",
       "481     {'ID': 'AT1G01070.2', 'Parent': 'AT1G01070', '...  \n",
       "593     {'ID': 'AT1G01110.1', 'Parent': 'AT1G01110', '...  \n",
       "620     {'ID': 'AT1G01120.1', 'Parent': 'AT1G01120', '...  \n",
       "820     {'ID': 'AT1G01180.1', 'Parent': 'AT1G01180', '...  \n",
       "846     {'ID': 'AT1G01200.1', 'Parent': 'AT1G01200', '...  \n",
       "...                                                   ...  \n",
       "927130  {'ID': 'ATCG01230.1', 'Parent': 'ATCG01230', '...  \n",
       "927137  {'ID': 'ATCG01240.1', 'Parent': 'ATCG01240', '...  \n",
       "927152  {'ID': 'ATCG01270.1', 'Parent': 'ATCG01270', '...  \n",
       "927157  {'ID': 'ATCG01280.1', 'Parent': 'ATCG01280', '...  \n",
       "927165  {'ID': 'ATCG01300.1', 'Parent': 'ATCG01300', '...  \n",
       "\n",
       "[8894 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair_df[tair_df.features_dict.apply(lambda x: x.get('conf_rating', '') == '*****')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5ba844e",
   "metadata": {},
   "source": [
    "# Filter TAIR entries based on feature(s)\n",
    "\n",
    "#### Filtering entries that are protein coding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f822765f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>data_source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>features</th>\n",
       "      <th>features_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>3631.0</td>\n",
       "      <td>5899.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01010;Name=AT1G01010;Note=NAC domain co...</td>\n",
       "      <td>{'ID': 'AT1G01010', 'Name': 'AT1G01010', 'Note...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>6788.0</td>\n",
       "      <td>9130.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01020;Name=AT1G01020;Note=Arv1-like pro...</td>\n",
       "      <td>{'ID': 'AT1G01020', 'Name': 'AT1G01020', 'Note...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>11649.0</td>\n",
       "      <td>13714.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01030;Name=AT1G01030;Note=AP2/B3-like t...</td>\n",
       "      <td>{'ID': 'AT1G01030', 'Name': 'AT1G01030', 'Note...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>23121.0</td>\n",
       "      <td>31227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01040;Name=AT1G01040;Note=dicer-like 1;...</td>\n",
       "      <td>{'ID': 'AT1G01040', 'Name': 'AT1G01040', 'Note...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>31170.0</td>\n",
       "      <td>33171.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=AT1G01050;Name=AT1G01050;Note=pyrophosphory...</td>\n",
       "      <td>{'ID': 'AT1G01050', 'Name': 'AT1G01050', 'Note...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome data_source  type    start      end score strand frame  \\\n",
       "0         Chr1   Araport11  gene   3631.0   5899.0     .      +     .   \n",
       "17        Chr1   Araport11  gene   6788.0   9130.0     .      -     .   \n",
       "138       Chr1   Araport11  gene  11649.0  13714.0     .      -     .   \n",
       "157       Chr1   Araport11  gene  23121.0  31227.0     .      +     .   \n",
       "249       Chr1   Araport11  gene  31170.0  33171.0     .      -     .   \n",
       "\n",
       "                                              features  \\\n",
       "0    ID=AT1G01010;Name=AT1G01010;Note=NAC domain co...   \n",
       "17   ID=AT1G01020;Name=AT1G01020;Note=Arv1-like pro...   \n",
       "138  ID=AT1G01030;Name=AT1G01030;Note=AP2/B3-like t...   \n",
       "157  ID=AT1G01040;Name=AT1G01040;Note=dicer-like 1;...   \n",
       "249  ID=AT1G01050;Name=AT1G01050;Note=pyrophosphory...   \n",
       "\n",
       "                                         features_dict  \n",
       "0    {'ID': 'AT1G01010', 'Name': 'AT1G01010', 'Note...  \n",
       "17   {'ID': 'AT1G01020', 'Name': 'AT1G01020', 'Note...  \n",
       "138  {'ID': 'AT1G01030', 'Name': 'AT1G01030', 'Note...  \n",
       "157  {'ID': 'AT1G01040', 'Name': 'AT1G01040', 'Note...  \n",
       "249  {'ID': 'AT1G01050', 'Name': 'AT1G01050', 'Note...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair_df[tair_df.features_dict.apply(lambda x: x.get('locus_type', '') == 'protein_coding')].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ecd732a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ID=AT1G01010;Name=AT1G01010;Note=NAC domain containing protein 1;symbol=NAC001;full_name=NAC domain containing protein 1;computational_description=NAC domain containing protein 1;locus=2200935;locus_type=protein_coding'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tair_df.features.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89653cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bracu",
   "language": "python",
   "name": "bracu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
